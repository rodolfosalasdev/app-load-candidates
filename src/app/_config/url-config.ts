export const urlConfig = Object.freeze({
    createCandidateUrl: 'http://localhost:3000/candidates/create-candidate'
})