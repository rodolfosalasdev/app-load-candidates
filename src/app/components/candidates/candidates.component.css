.candidates {
    display: flex;
    flex-direction: column;
    justify-content: center;
    gap: 2rem;
    width: 100%;
}