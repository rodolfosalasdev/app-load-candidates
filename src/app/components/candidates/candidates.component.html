<div class="candidates">
    <app-form-candidates 
        (formOutput)="formValidate($event)"
        (fileOutput)="processExcelFile($event)"
    />
    <app-candidates-list />
</div>
