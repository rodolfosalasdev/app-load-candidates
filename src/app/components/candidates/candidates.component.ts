import { Component, inject } from '@angular/core';
import { FormCandidatesComponent } from './form-candidates/form-candidates.component';
import { CandidatesListComponent } from './condidates-list/candidates-list.component';
import { FormGroup } from '@angular/forms';
import * as XLSX from 'xlsx';
import { ICreateCandidate } from '../../_interfaces/create-candidate.interface';
import { CreateCandidateService } from '../../_services/create-candidate.service';

@Component({
  selector: 'app-candidates',
  standalone: true,
  imports: [FormCandidatesComponent, CandidatesListComponent],
  templateUrl: './candidates.component.html',
  styleUrl: './candidates.component.css'
})
export class CandidatesComponent {
  private service = inject(CreateCandidateService);
  private file!: any
  private params: ICreateCandidate = {
    name: '',
    surname: '',
    file: {
      seniority: '',
      years: 0,
      availability: false,
    },
  };

  public formValidate(form: FormGroup) {
    if (form.valid) {
      this.params.name = form.get('name')?.value || '';
      this.params.surname = form.get('surname')?.value || '';
      this.createCandidates();
    } else {
      console.error('Form is invalid');
    }
  }

  private createCandidates(): void {
    this.service.createCandidate(this.params).subscribe({
      next: () => {
        this.service.getCandidates();
        this.file = null;
        this.params.file.availability = false;
      },
      error: (error) => {
        console.error('Error creating candidate:', error);
      },
    });
  }

  public processExcelFile(file: File): void {
    this.file = file;
    const reader = new FileReader();
    reader.onload = (e) => {
      const data = new Uint8Array(e.target?.result as ArrayBuffer);
      const workbook = XLSX.read(data, { type: 'array' });
      const worksheet = workbook.Sheets[workbook.SheetNames[0]];
      const excelData = XLSX.utils.sheet_to_json(worksheet, { header: 1 });

      this.extractExcelData(excelData);
    };
    reader.readAsArrayBuffer(this.file);
  }

  private extractExcelData(excelData: any[]): void {
    excelData.forEach((row: any) => {
      if (row[1]) {
        const value = row[1];
        if (typeof value === 'string') this.params.file.seniority = value;
        if (typeof value === 'number') this.params.file.years = value;
        if (typeof value === 'boolean') this.params.file.availability = value;
      }
    });
  }

}
