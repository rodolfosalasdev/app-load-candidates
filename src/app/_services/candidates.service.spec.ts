import { TestBed } from '@angular/core/testing';

import { CandidatesService } from './candidates.service';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { ICandidatesListResponse } from '../_interfaces/candidates-liste-response.interface';
import { urlConfig } from '../_config/url-config';
import { ICreateCandidate } from '../_interfaces/create-candidate.interface';


describe('CandidatesService', () => {
  let service: CandidatesService;
  let httpMock: HttpTestingController;

  beforeEach(async() => {
    await TestBed.configureTestingModule({
      imports: [HttpClientTestingModule]
    });
    service = TestBed.inject(CandidatesService);
    httpMock = TestBed.inject(HttpTestingController);
  });

  afterEach(() => {
    httpMock.verify();
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  it('deve buscar os candidatos e atualizar o signal', () => {
    const mockCandidates: ICandidatesListResponse[] = [
      {
        name: 'Rodolfo',
        surname: 'Salas',
        seniority: 'Senior',
        years: 5,
        availability: true
      },
      {
        name: 'Larissa',
        surname: 'Salas',
        seniority: 'Junior',
        years: 1,
        availability: false
      },
    ];

    // Chama o método que faz a requisição HTTP
    service.getCandidates();

    // Simula a resposta da chamada GET
    const req = httpMock.expectOne(urlConfig.candidatesListUrl);
    expect(req.request.method).toBe('GET');
    req.flush(mockCandidates); // Retorna os dados simulados

    // Verifica se o signal foi atualizado corretamente
    expect(service.candidatesSignal()).toEqual(mockCandidates);
  });

  it('deve criar um candidato via POST', () => {
    const mockCandidate: ICreateCandidate = { 
      name: 'Rodolfo',
      surname: 'Salas',
      file: {
        seniority: 'Senior',
        years: 5,
        availability: true
      }
    };

    // Chama o método que faz a requisição POST
    service.createCandidate(mockCandidate).subscribe((candidate) => {
      expect(candidate).toEqual(mockCandidate);
    });

    // Simula a resposta da chamada POST
    const req = httpMock.expectOne(urlConfig.createCandidateUrl);
    expect(req.request.method).toBe('POST');
    expect(req.request.body).toEqual(mockCandidate);
    req.flush(mockCandidate); // Retorna os dados simulados
  });
});
