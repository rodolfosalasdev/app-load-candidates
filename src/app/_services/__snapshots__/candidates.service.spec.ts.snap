// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CandidatesService should create a candidate POST 1`] = `"POST"`;

exports[`CandidatesService should create a candidate POST 2`] = `
{
  "file": {
    "availability": true,
    "seniority": "Senior",
    "years": 5,
  },
  "name": "Rodolfo",
  "surname": "Salas",
}
`;

exports[`CandidatesService should create a candidate POST 3`] = `
{
  "file": {
    "availability": true,
    "seniority": "Senior",
    "years": 5,
  },
  "name": "Rodolfo",
  "surname": "Salas",
}
`;

exports[`CandidatesService should search for candidates and update the signal 1`] = `"GET"`;

exports[`CandidatesService should search for candidates and update the signal 2`] = `
[
  {
    "availability": true,
    "name": "Rodolfo",
    "seniority": "Senior",
    "surname": "Salas",
    "years": 5,
  },
  {
    "availability": false,
    "name": "Larissa",
    "seniority": "Junior",
    "surname": "Salas",
    "years": 1,
  },
]
`;
